mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -0,0 +1,310 @@
+                                                 /*outputs=*/qregType);
+    func::FuncOp fnAlloc = rewriter.create<func::FuncOp>(loc, fnAllocName, fnAllocType);
+    fnAlloc.setPrivate();
+    Block *allocBloc = fnAlloc.addEntryBlock();
+    rewriter.setInsertionPointToStart(allocBloc);
+    Value nQubits = allocBloc->getArgument(0);
+    IntegerAttr intAttr;
+    auto qreg = rewriter.create<quantum::AllocOp>(loc, qregType, nQubits, intAttr);
+    rewriter.create<func::ReturnOp>(loc, qreg.getResult());
+    return SymbolRefAttr::get(ctx, fnAllocName);
+}
+FlatSymbolRefAttr ZneLowering::getOrInsertFnWithoutMeasurements(Location loc,
+                                                                PatternRewriter &rewriter,
+                                                                mitigation::ZneOp op)
+{
+    MLIRContext *ctx = rewriter.getContext();
+    OpBuilder::InsertionGuard guard(rewriter);
+    ModuleOp moduleOp = op->getParentOfType<ModuleOp>();
+    Type qregType = quantum::QuregType::get(rewriter.getContext());
+    func::FuncOp fnOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+    TypeRange originalTypes = op.getArgs().getTypes();
+
+    std::string fnWithoutMeasurementsName = op.getCallee().str() + ".withoutMeasurements";
+    if (moduleOp.lookupSymbol<func::FuncOp>(fnWithoutMeasurementsName)) {
+        return SymbolRefAttr::get(ctx, fnWithoutMeasurementsName);
+    }
+
+    SmallVector<Type> typesWithoutMeasurements = {originalTypes.begin(), originalTypes.end()};
+    typesWithoutMeasurements.push_back(qregType);
+
+    FunctionType fnWithoutMeasurementsType = FunctionType::get(ctx, /*inputs=*/
+                                                               typesWithoutMeasurements,
+                                                               /*outputs=*/qregType);
+    func::FuncOp fnWithoutMeasurementsOp =
+        rewriter.create<func::FuncOp>(loc, fnWithoutMeasurementsName, fnWithoutMeasurementsType);
+    fnWithoutMeasurementsOp.setPrivate();
+    rewriter.cloneRegionBefore(fnOp.getBody(), fnWithoutMeasurementsOp.getBody(),
+                               fnWithoutMeasurementsOp.end());
+    Block *fnWithoutMeasurementsBlock = &fnWithoutMeasurementsOp.front();
+    fnWithoutMeasurementsBlock->addArgument(qregType, loc);
+    quantum::AllocOp allocOp = *fnWithoutMeasurementsOp.getOps<quantum::AllocOp>().begin();
+
+    auto lastArgIndex = fnWithoutMeasurementsBlock->getArguments().size();
+    allocOp.replaceAllUsesWith(fnWithoutMeasurementsBlock->getArgument(lastArgIndex - 1));
+
+    rewriter.eraseOp(allocOp);
+    rewriter.setInsertionPointToStart(&fnWithoutMeasurementsOp.getBody().front());
+
+    std::vector<Operation *> insertOps;
+    fnWithoutMeasurementsOp.walk(
+        [&](quantum::InsertOp insertOp) { insertOps.push_back(insertOp); });
+    fnWithoutMeasurementsOp.walk(
+        [&](func::ReturnOp returnOp) { returnOp->setOperands(insertOps.back()->getResult(0)); });
```

Comment:
I don't get why do we need this `insertOps` while the last Op is only required.  :thinking: Can it be replaced with `Operation * lastOp`?

Reply:
I think here it is guaranteed to have single last insert operation. Is there another way to get the last one?

Reply:
What about the following suggestion?

```
    Operation * lastOp;
    fnWithoutMeasurementsOp.walk(
        [&](quantum::InsertOp insertOp) { lastOp = insertOp; });
    fnWithoutMeasurementsOp.walk(
        [&](func::ReturnOp returnOp) { returnOp->setOperands(lastOp->getResult(0)); });
```

Reply:
Thanks added!